src/Streamly/Internal/Data/Array/Foreign/Type.hs

```diff
@@ -158,6 +159,13 @@ data Array a =
 foreign import ccall unsafe "string.h strlen" c_strlen
     :: CString -> IO CSize
 
+-- | Similar to unsafeWithForeignPtr.
+unsafeWithArray :: MonadIO m => Array a -> (Ptr a -> m b) -> m b
+unsafeWithArray Array{..} f = do
+  r <- f arrStart
+  liftIO $ touch arrContents
+  return r
```

Comment:
This should be `unsafeAsPtr`, I know there is a cast in unsafeAsPtr but the users of the API can do the cast themselves when needed.